src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,78 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private Long maxId;
+    private List<Product> products;
+    private final Object lock = new Object();
+
+    public ArrayListProductDao() {
+        products = new ArrayList<>();
+        maxId = 0L;
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws ProductNotFoundException {
```

Comment:
Please add null parameter handling for this class. Follow my recommendations from our call

Reply:
Also please add synchronization whenever shared state in this class accessed (for writing or reading)